Add logic for initializing CSINodeInfo on Node startup

davidz627 opened the pull request:
This PR adds the logic for initializing CSINodeInfo with well-known drivers used for migration during node startup. This is necessary so that the Attach Detach controller can see the availability/pluginmechanism matrix when making decisions about how to deal with volumes.

/kind feature

/assign @saad-ali @verult @jsafrane 
/cc @leakingtapan @ddebroy 

Fixes #70514

```release-note
If CSINodeInfo and CSIMigration are active in the cluster, Kubelet will post NotReady until CSINodeInfo CRD is installed and initialized with basic volume plugin mechanism information for well-known drivers
```

davidz627 commented:
/assign @liggitt

davidz627 commented:
This PR will target 1.14 and will probably contain logic to post Kubelet `NotReady` if CSINodeInfo CRD is not installed

davidz627 committed:
Block Kubelet NodeReady on CSINodeInfo install and initialization if CSINodeInfo and CSIMigration are on

davidz627 commented:
This PR is ready for review, it is targeted for k8s v1.14

/assign @jsafrane @saad-ali @leakingtapan @yujuhong 
/cc @ddebroy @msau42

liggitt commented on pkg/controller/volume/attachdetach/attach_detach_controller.go:
```diff
@@ -659,6 +659,10 @@ func (adc *attachDetachController) GetPluginDir(podUID string) string {
 	return ""
 }
 
+func (adc *attachDetachController) SetKubeletError(err error) {
```
having to add this method to unrelated controllers indicates the method probably doesn't belong in the interface it was added to

davidz627 replied:
this is an unfortunate remnant of the `VolumeHost` structure. If you look at all the other methods in volume host they are exactly the same, in reality volumehost should *just* be a Kubelet concept but the other volume controllers kind of ended up using it too so they all just implement nil/empty methods. It is a known issue and is due for a refactor but that is outside of the scope of this PR.

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +248,84 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+
+	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host, migratedDrivers)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) &&
```
is it valid to enable CSIMigration without CSINodeInfo? does anything enforce that?

davidz627 replied:
I don't think anything enforces that, is there a way to specify dependencies between feature flags?

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +248,84 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		// TODO(leakingtpan): Add AWS migration feature gates and place them here
+	}
+
+	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host, migratedDrivers)
```
should this only be initialized if CSINodeInfo is enabled?

assigning to a global here is a code smell. is there a reason not to put these in the csiPlugin instance?

davidz627 replied:
1. probably yes, I will gate it.

2. Not sure, I didn't actually write this code I just added the `migratedDrivers` parameter. Any bigger changes are outside the scope of this PR as I'm not sure what side effects they may have
/cc @verult 
who I believe is familiar with (or wrote) this code

hoegaarden replied:
@mariantalla and /me are planning to address the globals in `pkg/volume/csi/...`, so I suggest to leave it as is for now.

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +248,84 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
 
+func initializeCSINodeInfo(host volume.VolumeHost) {
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil || len(host.GetPodsDir()) == 0 {
```
is `len(host.GetPodsDir()) == 0` the canonical way to check if we are not on the kubelet? that seems rather indirect

davidz627 replied:
I don't know about "canonical" since I don't see anyone else doing it. But from what I understand these indirect checks are the "best" way right now, see comment: https://github.com/kubernetes/kubernetes/pull/70909#discussion_r242768000

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,13 +92,124 @@ type Interface interface {
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		klog.Warningf("InitializeMigratedDrivers called but CSINodeInfo or CSIMigration is disabled")
```
why is this not an error case?

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,13 +92,124 @@ type Interface interface {
+			nodeInfo, err = nim.CreateCSINodeInfo()
+		}
+		if err != nil {
+			updateErrs = append(updateErrs, err)
```
this errors list will grow without bound if the create call perpetually fails, right?

liggitt replied:
if the create call returns a NotFound error, that means the API is not available. do we want to keep retrying in that case?

davidz627 replied:
yes, it could grow without bound (assuming we retry indefinitely like you suggested in another comment). What is the canonical way of dealing with this? It doesn't seem like we can have it both ways (have an error array with useful information, and have it not grow forever) unless we have some extra logic maybe to ignore duplicate errors?

We want to keep retrying forever, Kubelet should not/can not start without this.

liggitt replied:
the call site that calls this also retries... retries inside retries get really tricky to reason about, and we probably want to surface the error via node status. can we drop the retry in InitializeMigratedDrivers(), return errors if they are encountered, and count on the caller to re-call InitializeMigratedDrivers()?

liggitt commented on pkg/volume/plugins.go:
```diff
@@ -367,6 +367,9 @@ type VolumeHost interface {
 	GetEventRecorder() record.EventRecorder
+
+	// SetKubeletError sets the ready state on the kubelet.
+	SetKubeletError(err error)
```
can't this already be accomplished using GetKubeClient() and GetNodeName()? if so, let's make a utility method that does the setting rather than adding to this interface and making unrelated objects implement empty functions

davidz627 replied:
I'm not familiar with the method you're referring to here. How do I set Kubelet ready status using KubeClient and NodeName, you mean by updating the API object directly? will Kubelet not then overwrite the ready status next time it polls the readystatusfuncs?

liggitt replied:
yeah, read further and figured out what this is for. made a suggestion earlier about a kubelet-specific volume host interface

davidz627 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -74,11 +80,108 @@ type Interface interface {
+		}
+
+		newDriverSpecs = append(newDriverSpecs, driverSpec)
+		newDriverStatuses = append(newDriverStatuses, driverStatus)
```
As per offline discussion today:

My plan is to continue to work on it and get it in with some form of the current "updating spec and status together" code smell. With a big TODO/Issue linked so that it will all get addressed in a Q1 refactor which will introduce the reconciler concept to all the pieces of the nodeinfomanager. However, I will update the code to make two calls sequentially (one to the main resource for spec, one to the subresource for status) so I can shape the API now then split the code internally into reconcile loops later

davidz627 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +248,84 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+				}
+
+				for _, g := range groupResource.APIResources {
+					if g.Name == "csinodeinfos" {
```
It could, but I think it is outside the scope of this PR

davidz627 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +248,84 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		found := false
+		for {
+			select {
+			case <-crdTicker.C:
```
changing to exponential backoff model

davidz627 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,13 +92,124 @@ type Interface interface {
+	for driver, _ := range nim.migratedDrivers {
+		if existingDrivers.Has(driver) {
+			// Skip adding new driver because already updated previously
+			continue
```
1. Spec and status will eventually be modified in two seperate reconcile loops, spec will be reconciled from existance of driver sockets, and status will be reconciled based on availability of driver, though this has not been implemented yet. We will probably scope it for Q2.

2. I believe what you're referring to here is a reconciler style nodeinfomanager where the "authoritative database" would be the sockets in the plugin registration folder. Let me know if I misunderstood.

Kubelet should definitely not give up saving CSINodeInfo, I can fix that in this patch.

I believe there are no races between `tryInitializeMigratedDriver` and dynamic driver registration, since both of them retry on update conflict neither will stomp on the other changes. dynamic driver registration is considered authoritative, and `tryInitializeMigratedDriver` intention is to maintain the drivers registered dynamically as-is (maybe update VolumePluginMechanism) and only add new drivers for those that do not exist in CSINodeInfo thus far. I have test and verify this behavior in `nodeinfomanager_test.go`, if the code does not seem to be doing the right thing please point out specifics in the next round of review.

Thanks for the thorough review

davidz627 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,13 +92,124 @@ type Interface interface {
+		driverStatus := csiv1alpha1.CSIDriverInfoStatus{
+			Name: driver,
+			// We are just adding this driver object it's not installed
+			Available:             false,
```
See second last paragraph here addressing this comment
https://github.com/kubernetes/kubernetes/pull/70909#discussion_r242767419

davidz627 committed:
Split InitializeMigratedDrivers into Spec and Status loops in prep for reconciliation refactor in Q2. Addressed various other comments.

davidz627 commented:
@liggitt @jsafrane 
I "resolved" comments that I addressed in code. Other comments I have replied to with either explanations or further questions. Thanks for your reviews.

Please take another look. Thanks!

jsafrane commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -441,6 +597,13 @@ func (nim *nodeInfoManager) installDriverToCSINodeInfo(
 		topologyKeys.Insert(k)
 	}
 
+	volumePluginMechanism := csiv1alpha1.VolumePluginMechanismInTree
```
Reuse getVolumePluginMechanism?

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	}
+
+	go func() {
+		host.SetKubeletError(fmt.Errorf("Could not find CSINodeInfo CRD. Please install the CSINodeInfo CRD"))
```
setting the error inside the gofunc is racy... need to initialize it synchronously before starting the gofunc (and start it at something more accurate like "CSINodeInfo availability not yet verified"). don't include "CRD" in the error message

davidz627 replied:
I tried doing this earlier, however the issue then is then since clearing the error is the job of the CSI Volume Plugin, if the user is to turn CSI off, the error may not be cleared...

I guess it could clear it inside the init. And also there should be some logic outside of volume plugin initialization that clears it if CSI is off in general. How does that sound?

liggitt replied:
the only way to turn it off is via the feature gate, right? doesn't that requires a restart of the kubelet, which would mean initializeCSINodeInfo() would not get called on startup, and the error would remain nil as it is on startup

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+			err := wait.ExponentialBackoff(crdBackoff, func() (bool, error) {
+				groupResource, err := kubeClient.Discovery().ServerResourcesForGroupVersion("csi.storage.k8s.io" + "/" + "v1alpha1")
+				if err != nil {
+					klog.Warningf("Failed to get groupResource: %v", err)
```
can update storage error to something informative here

liggitt replied:
and should this be logged as an error?

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+			}
+		}
+
+		host.SetKubeletError(fmt.Errorf("Found CSINodeInfo CRD. Waiting for CSINodeInfo initialization"))
```
suggest "CSINodeInfo not yet initialized"

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+				klog.V(4).Infof("Intializing migrated drivers on CSINodeInfo")
+				err := nim.InitializeMigratedDrivers()
+				if err != nil {
+					klog.Warningf("Failed to initialize migrated drivers: %v", err)
```
update the storage error to surface errors here?

liggitt replied:
and should this be logged as an error?

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,161 @@ type Interface interface {
+
+		driverSpec := csiv1alpha1.CSIDriverInfoSpec{
+			Name:         driver,
+			NodeID:       "",
```
what are observers told to do with an empty value? does that mean "this is not yet populated" or "use the node name as the node ID"?

davidz627 replied:
It means "this is not yet populated". At the time of initialization if the driver does not exist we have no way of getting that information.

liggitt replied:
that's fine, as long as everything making use of this field has the same understanding, and waits for the field to be populated rather than assuming an empty field means to use some other value as a fallback

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,161 @@ type Interface interface {
+	// TODO(TODO) We can remove this once nodeinfomanager is refactored into a reconcile pattern because
+	// then we can just add the specs in the above loop and the reconciler will fill in the statuses based
+	// on that.
+	err = wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
```
same comment here about not retrying inside InitializeMigratedDrivers and depending on the caller to re-call us

liggitt replied:
can also use the updated object that was the output of the previous step instead of looking it up again

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,161 @@ type Interface interface {
+			TopologyKeys: []string{},
+		}
+
+		newDriverSpecs = append(newDriverSpecs, driverSpec)
```
keep track of whether we needed to do this... if not, skip the update

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+			Factor:   5.0,
+			Jitter:   0.1,
+		}
+		for {
```
combining `for` + `wait.ExponentialBackoff` has an odd burst pattern, where the attempts backoff until they reach the max exponential steps, then exit the ExponentialBackoff with an error, then restart the for loop fast again.

might be clearer to do something like:
```go
backoff := &wait.Backoff{
  Steps: 5,
  Duration: 10 * time.Millisecond,
  Factor: 4.0,
  Jitter: 0.1,
  Cap: 10 * time.Second,
}
for {
  time.Sleep(backoff.Step())
  groupResource, err := kubeClient.Discovery().ServerResourcesForGroupVersion("csi.storage.k8s.io" + "/" + "v1alpha1")
  if err != nil {
    ... log, report, continue ...
  }
  ... look for resource, if missing, log, report, continue
  break
}
```

that lets the backoff grow and settle at 10 second interval checks

same comment applies to the for/ExponentialBackoff below. would need to reset the Backoff struct

liggitt replied:
an eventual "give up, go out, and come back in again" restart is probably better. I'm not sure what the other init-time checks do w.r.t. specific timeframes. Would recommend asking @kubernetes/sig-node-pr-reviews about what other subsystems (network, container runtime, etc) do

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,161 @@ type Interface interface {
+		driverSpec := csiv1alpha1.CSIDriverInfoSpec{
+			Name:         driver,
+			NodeID:       "",
+			TopologyKeys: []string{},
```
same question for topology keys... if an observer sees this before it gets populated, would it cause it to do something incorrect, or to wait?

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,161 @@ type Interface interface {
+			Available:             false,
+			VolumePluginMechanism: volumePluginMechanism,
+		}
+		newDriverStatuses = append(newDriverStatuses, driverStatus)
```
track whether we needed to do this, and if we didn't append new statuses, skip the updatestatus

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,161 @@ type Interface interface {
+	}
+
+	nodeInfo.Status.Drivers = newDriverStatuses
+	_, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Update(nodeInfo)
```
add a TODO to call UpdateStatus() in v1beta1

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -441,6 +597,13 @@ func (nim *nodeInfoManager) installDriverToCSINodeInfo(
+	}
+
 	specModified := true
 	statusModified := true
```
should these start false and be turned true if we append or modify anything in the loops?

davidz627 replied:
since we're only modifying one driver here it doesn't really matter which way we do it. starting `true` and flipping to `false` seems to make the internal `if existingDriver == newDriver` checks a lot cleaner

liggitt commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -460,8 +623,7 @@ func (nim *nodeInfoManager) installDriverToCSINodeInfo(
-				driverInfoStatus.VolumePluginMechanism == csiv1alpha1.VolumePluginMechanismInTree {
+				driverInfoStatus.VolumePluginMechanism == volumePluginMechanism {
 				statusModified = false
 			}
```
should there be an `else` block here that does this?
```
driverInfoStatus.Available = true
driverInfoStatus.VolumePluginMechanism = volumePluginMechanism
newDriverStatuses = append(newDriverStatuses, driverInfoStatus)
statusModified = true
```

that would preserve the status order, which is nice

liggitt replied:
same question for spec

davidz627 replied:
ah I see what you're saying. I can clean this up a bit

davidz627 replied:
I changed it but ended up reverting. The amount of flags made it a little confusing. There was "statusModified" and "specModified" as well as "statusExists" and "specExists" since we need to cover the case where the driver is not currently represented in CSINodeInfo at all.

liggitt replied:
fair enough

liggitt commented:
`verify no-vendor-cycles` is unhappy with pulling in an out-of-tree repo (github.com/kubernetes-csi/kubernetes-csi-migration-library) that depends on staging repos. I assume that's what the [request](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-architecture/7LqAhO_RPaI/4Ks2HvmuEQAJ) for k8s.io/csi-translation-lib was for, and the code would be moved there?

liggitt commented:
some gofmt/lint errors as well

ddebroy commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -44,6 +44,8 @@ import (
 	"k8s.io/kubernetes/pkg/volume"
 	"k8s.io/kubernetes/pkg/volume/csi/nodeinfomanager"
+
+	csiPlugins "github.com/kubernetes-csi/kubernetes-csi-migration-library/plugins"
```
Yes that is correct. This is the updated thread: https://groups.google.com/d/msg/kubernetes-sig-architecture/yYcSj75gaGo/Nfd0g373FQAJ

davidz627 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		// This function prevents Kubelet from posting Ready status until CSINodeInfo
+		// is both installed and initialized
+		initializeCSINodeInfo(host)
```
yes, but even if it is more I believe 2 of these could run in parallel without issue.

liggitt replied:
initializeCSINodeInfo() sets and clears the error status... if instances were getting created at times other than startup, they could affect the node's ready status

davidz627 committed:
Review changes #2

davidz627 commented:
new commit addressing review comments. will fix vendor cycles issue when `csi-translation-lib` populated: https://github.com/kubernetes/kubernetes/pull/72770

davidz627 commented:
Hi @kubernetes/sig-node-pr-reviews,
this PR is ready for review by your SIG. It has significant implications for Kubelet startup when the `CSIMigration` Feature Flags are on (which we plan on turning on by default in `1.15`) but are off by default in `1.14`. The general idea is that we block Kubelet ready status on existence of `CSINodeInfo` as well as initialization of `CSINodeInfo` as going forward these will be necessary prerequisites to the functionality of the storage subsystem in Kubernetes.

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -263,72 +265,89 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	}
 	kubeClient := host.GetKubeClient()
-	if kubeClient == nil || len(host.GetPodsDir()) == 0 {
+	if kubeClient == nil {
```
this will be nil on a kubelet that is running without talking to an apiserver (without --kubeconfig provided). do we avoid calling initializeCSINodeInfo in that scenario?

davidz627 replied:
yes, I believe we can't populate an API object with data without talking to an apiserver

liggitt replied:
my question is if this code path (initializing a CSI storage driver) is possible to exercise in a kubelet that is not running against an API server. If so, this would seem to make CSI drivers impossible to initialize successfully, and would make a kubelet like that crashloop, right?

liggitt replied:
if we intentionally don't have a connection to an API server, should the CSI driver report initialization success so the kubelet can start up and run on its own successfully?

yujuhong replied:
+1 wouldn't this cause standalone kubelets to fail?

What's the expected behavior for a standalone kubelet? Will it still be able to attach and mount PDs when `EnableControllerAttachDetach` is set to false?

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -263,72 +265,89 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		})
+		if err != nil {
+			// Kill the Kubelet process and allow it to restart to retry initialization
+			os.Exit(1)
```
I expected an error propagation, not an os.Exit(1) buried deep in init... do we have other examples of init loops in the kubelet triggering exits? if we can't propgate an error, at least exit via `klog.Fatalf` so we get a message indicate what caused the exit

davidz627 replied:
since this is inside a `go func` I don't think error propagating is an option here. `klog.Fatalf` seems like the right choice

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -263,72 +265,89 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+			Jitter:   0.1,
+		}
+		err = wait.ExponentialBackoff(initBackoff, func() (bool, error) {
+			klog.V(4).Infof("Intializing migrated drivers on CSINodeInfo")
```
typo: "Initializing"

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -263,72 +265,89 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		})
+		if err != nil {
+			// Kill the Kubelet process and allow it to restart to retry initialization
+			os.Exit(1)
```
same comment about propagating or log.Fatal

davidz627 commented:
/retest

liggitt commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	go func() {
+		defer utilruntime.HandleCrash()
+
+		// Backoff parameters tuned to retry over 140 seconds. Will fail and restart the Kubelet
```
cc @smarterclayton @kubernetes/sig-cluster-lifecycle for possible implications on self-hosted cases... how long should the api-contacting portion of a kubelet wait before giving up and triggering kubelet exit?

neolit123 replied:
/assign @timothysc 

from my perspective it seems wrong for a CSI plugin or any plugable solution to manage a kublet restart in the first place. the kubelet is a system service.
kubelet restarts should be managed by higher level tools that the user invokes close to the OS.

i'm finding it difficult understanding why this is required from `csi_plugin.go`?

> how long should the api-contacting portion of a kubelet wait before giving up and triggering kubelet exit?

kubeadm is doing 40seconds for the healthz endpoint check:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/apiclient/wait.go#L164

davidz627 replied:
`csi_plugin.go` is not the pluggable solution itself. It is the system piece that will connect to the various CSI Drivers that are eventually plugged in to Kubernetes.

This init step will be *required* to "boot the volume subsystem" for future versions of Kubernetes.

neolit123 replied:
> kubeadm is doing 40seconds for the healthz endpoint check:

correction. it's 155 seconds, but there was an initial timeout of 40 seconds.

> csi_plugin.go is not the pluggable solution itself. It is the system piece that will connect to the various CSI Drivers that are eventually plugged in to Kubernetes.

i guess i was misled by the file name, as it implies a plugin.

yujuhong replied:
Currently, kubelet doesn't fail if it has problems reaching the apisver. Maybe we should keep that behavior unless there is clear benefit of restarting kubelet.

neolit123 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		return fmt.Errorf("failed to get KubeClient")
+	}
+
+	kvh.SetKubeletError(fmt.Errorf("CSINodeInfo is not yet available"))
```
errors without arguments should be created with `errors.New`.
applies to multiple lines in the diff.
[1]

neolit123 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil {
+		// If we're NOT on the Kubelet OR KubeClient DOESNT exist
+		return fmt.Errorf("failed to get KubeClient")
```
[1]

neolit123 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+			klog.Fatalf("Failed to find CSINodeInfo on API Server after retrying")
+		}
 
+		kvh.SetKubeletError(fmt.Errorf("CSINodeInfo available. CSINodeInfo not yet initialized"))
```
[1]

neolit123 commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		})
+		if err != nil {
+			// Kill the Kubelet process and allow it to restart to retry initialization
+			klog.Fatalf("Failed to initialize CSINodeInfo after retrying")
```
i though that usage of `klog.Fatalf` was discouraged, since if someone vendors such a package their app might exit in a unwanted manner.

also calling exiting from plugin seems quite wrong, in general.

neolit123 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,158 @@ type Interface interface {
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		return fmt.Errorf("initializeMigratedDrivers called but CSINodeInfo or CSIMigration is disabled")
```
[1]

neolit123 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,158 @@ type Interface interface {
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
```
[1]

neolit123 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,158 @@ type Interface interface {
+func (nim *nodeInfoManager) tryInitializeMigratedDriverSpecs(nodeInfo *csiv1alpha1.CSINodeInfo) (*csiv1alpha1.CSINodeInfo, error) {
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return nil, fmt.Errorf("error getting CSI client")
```
[1]

neolit123 commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,158 @@ type Interface interface {
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
```
[1]

yujuhong commented on pkg/volume/csi/nodeinfomanager/nodeinfomanager.go:
```diff
@@ -86,11 +92,158 @@ type Interface interface {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	// TODO(TODO): Change this into reconcile pattern
```
`TODO(TODO)`?

Is the TODO to fill in the name in the parentheses?

yujuhong commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+		return fmt.Errorf("failed to get KubeClient")
+	}
+
+	kvh.SetKubeletError(fmt.Errorf("CSINodeInfo is not yet available"))
```
Hmm....the storageError is not initialized to non-nil, so there could be a window where kubelet assumes storage is ready. You'll probably need to change that and also make sure standalone kubelet doesn't die because of this.

davidz627 replied:
i believe plugin init is done serially inside Kubelet init, so Kubelet does not even start posting ready status until after this has run. There wouldn't be any window where Kubelet assumes storage is ready when the Kubelet is actually running.

yujuhong commented on pkg/features/kube_features.go:
```diff
@@ -390,6 +390,18 @@ const (
+	// alpha: v1.14
+	//
+	// Enables the GCE PD in-tree driver to GCE CSI Driver migration feature.
+	CSIMigrationGCE utilfeature.Feature = "CSIMigrationGCE"
```
Is it necessary to have a CSIMigration feature, and features for individual drivers?

davidz627 replied:
yes, since we have to progress each plugin separately based on qualification. And we also want a way to turn the whole feature off if its broken

davidz627 commented:
/close
replaced by #74835
